### DISCUSSION TITLE
New version uncompleted?

### DISCUSSION THREAD

**Discussion by fus3** [fra589/cn5X#25] [General]
Hello Gauthier, i just downloaded the last version day (16/01/2021) and i jut tried it to run it from Visual Studio (because im adding a new macro button to probe the length of a new tool) but it shows some errors:

_PS C:\Users\paumo\Downloads\cn5X-master> & C:/Users/paumo/AppData/Local/Programs/Python/Python39/python.exe
c:/Users/paumo/Downloads/cn5X-master/cn5X-master/cn5X.py
Traceback (most recent call last):
 File "c:\Users\paumo\Downloads\cn5X-master\cn5X-master\cn5X.py", line 27, in <module>
    import serial, serial.tools.list_ports
ModuleNotFoundError: No module named 'serial'_

I see that you also added a new menu (dlgProbe.ui) for some probe actions, so i suppose that this version is not ready yet for using it, on the version 0.4.e. that i have on my raspberry pi 3 for my 5 axis cnc i can run it from Visual Studio without problems and add the new button. 

Also about the macro button that i added,

![TPbtn](https://user-images.githubusercontent.com/75036029/104820702-f6fa2900-5836-11eb-9f87-70269f894694.PNG)

The button works fine but when i press it seems that all the gcodes are send at the same time instead of doing one gcode at a time (like on a normal send cycle of a gcode file)

Thats the code of the button:

@pyqtSlot()
  def on_btnToolChangeHeightProbe(self):
    self.__grblCom.gcodeInsert("G90")
    self.__grblCom.gcodeInsert("G53 G1 A0 B0 F500")
    self.__grblCom.gcodeInsert("G53 G1 Z(distance of probe) F500")
    self.__grblCom.gcodeInsert("G53 G1 X(distance of probe) Y(distance of probe) F500")
    self.__grblCom.gcodeInsert("G91")
    self.__grblCom.gcodeInsert("G38.2 Z-100 F400")
    self.__grblCom.gcodeInsert("G1 Z5 F200")
    self.__grblCom.gcodeInsert("G38.2 Z-10 F40")
    self.__grblCom.gcodeInsert("G4 P.25")
    self.__grblCom.gcodeInsert("G38.4 Z10 F10")
    self.__grblCom.gcodeInsert("G4 P.25")
    self.__grblCom.gcodeInsert("G90")
    self.__grblCom.gcodeInsert("G10 L20 P1 Z(probe offset to WCS)")
    self.__grblCom.gcodeInsert("G91")
    self.__grblCom.gcodeInsert("G1 Z5 F200")
    self.__grblCom.gcodeInsert("G53 G1 Z-10 F500")
    self.__grblCom.gcodeInsert("G28 G91 Z0")
    self.__grblCom.gcodeInsert("G90")

**Comment by fra589** [comment]
Hi @fus3,  
  
Since version 0.5.a (Commits on Dec 30, 2020), QTSerialPort has been replaced by the pure python pyserial module which worked without (too much) problem with Microsoft Windows. This why you have the error ModuleNotFoundError.
You must install the python-serial module, you can use pip for this:
`pip install pyserial`

You are right, currently, I work on adding a new interface for the probing functionality.  
It will be able to probe tool length and made tool length correction automatically, but not only. It also be able to probe in any direction and define any origin position depending of probing, for example it will be capable of finding the center of a circle in XY plan and place the X and Y origin on this center point.  

The development of the program is a long process for me because I do not have enough time to devote to it, so, it will be necessary to be a little patient to have these complete functionalities.

@++;  
Gauthier.

**Comment by fra589** [comment]
Hi @fus3,

Discussion about probing here: https://github.com/fra589/cn5X/discussions/26

@++;
Gauthier.

**Reply by fra589** [reply]
New version 0.8 with probing functionality is available from today 😃

**Reply by fus3** [reply]
Wow, thats amazing!! I did test everything on my 3 axis (because my 5 Axis needs a preset WCS) and it works perfect!
